object/reply.go

```diff
@@ -65,6 +71,105 @@ func GetReplies(topicId int, memberId string, limit int, offset int) []*ReplyWit
 		v.Editable = isModerator || GetReplyEditableStatus(memberId, v.Author, v.CreatedTime)
 	}
 
+	var resultReplies []*ReplyWithAvatar
+
+	if enableCascadingReply {
+		replies = bulidReplies(replies)
+		//Use limit to calculate offset
+		//If limit is 2, but the first reply have 2 child replies(3 replies)
+		//We need put these replies to offset, so cannot use (page * limit) to calculate offset
+		pageLimit := limit
+		for index, reply := range replies {
+			replyLen := getReplyLen(reply)
+			//Ignore replies until page == 1
+			if page > 1 {
+				//Calculate limit in every ignore page
+				pageLimit -= replyLen
+				//Get replices for init == true(get the latest replies)
+				resultReplies = append(resultReplies, reply)
+				if pageLimit <= 0 {
+					page--
+					pageLimit = limit
+					if index + 1 < len(replies) {
+						//If the page is a usable value when we get the latest replies, clear the result
+						resultReplies = nil
+					}
+				}
+			} else if limit > 0 {
+				//if page == 1, prove that we are processing current page now
+				//So we can only calculate the limit and put replies to result slice
+				limit -= replyLen
+				resultReplies = append(resultReplies, reply)
+				page--
+			} else {
+				//if page == 1, and limit < 0, prove that we get all replies in this page now
+				break
+			}
+		}
+
+		if page > 0 {
+			realPage -= page
+		}
+	} else {
+		offset := page*limit - limit
+		for _, reply := range replies {
+			if offset > 0 {
+				offset--
+			} else {
+				if limit > 0 {
+					resultReplies = append(resultReplies, reply)
+				} else {
+					break
+				}
+			}
+		}
+	}
+
+	return resultReplies, realPage
+}
+
+func makeReplyTree(replies []*ReplyWithAvatar, reply *ReplyWithAvatar) bool {
+	if len(replies) == 0 {
+		return false
+	}
+	for _, r := range replies {
+		if r.Id == reply.ParentId {
+			r.Child = append(r.Child, reply)
+			return true
+		} else {
+			if makeReplyTree(r.Child, reply) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func getReplyLen(reply *ReplyWithAvatar) int {
```

Comment:
Maybe use a map to optimize this recursion function is better.
This makes too many duplicated calculate.